### PR TITLE
Fix behavior of BackdropFilter on a transparent background

### PR DESCRIPTION
**Pull request by flar**
This PR will need some tests, but I wanted to first submit it as a proposal for a new behavior for BackdropFilter.

Currently BackdropFilter will apply its blurred (or otherwise filtered) version of the background on top of the existing background with the default SrcOver blending mode. If the filter makes the image semi-transparent then that method of rendering the filtered background will leave parts of the original background visible through the semi-transparent areas.

This behavior has been noted by developers and considered a bug, as in

- https://github.com/flutter/flutter/issues/31706#issuecomment-522697161
- https://github.com/flutter/flutter/issues/60966

The blend mode which Skia uses to copy the filtered background back to the original surface can be controlled by the (optional) SkPaint object used when setting up the SaveLayer. If we set this blend mode to kSrc then the semi-transparent filtered backdrop will overwrite the original backdrop rather than merge with it. This behavior may be less confusing to developers.

I will need to write tests for this fix, but I wanted to propose it here as a WIP to get feedback on whether this behavior change is desirable or not for any reason including matching developer expectations.

The following dartpad entry shows the problem with using `BackdropFilter` to blur some text inside an `Opacity` widget. It also shows a workaround that places a solid white background behind the text inside the `Opacity` stack and another workaround using `ImageFiltered` instead of `BackdropFilter`. With the fix, all three would look similar (identical except for the content of the text being blurred).

https://dartpad.dev/a238a7c64d144ae4653faa7cc6d2ac4d

A screenshot showing the issue with `BackdropFilter` not obscuring the original text:

Screen Shot 2020-07-10 at 12 11 54 PM

### CODE REVIEW

**Comment by flar** [comment]
Pinging more developers to provide feedback on this change of behavior.

@renyou @mehmetf @goderbauer @Hixie @Piinks 

I haven't checked, but I don't believe this change would affect any existing golden tests as using a `BackdropFilter` on a non-opaque surface is a corner case that developers occasionally run into, but while using `BackdropFilter` in ways that I don't think it's original design was intended to support. Also, the recently added `ImageFiltered` widget is often a better tool to use for these particular uses of `BackdropFilter`.

**Comment by chinmaygarde** [comment]
@flar Removing the WIP tag as this seems to be waiting on feedback from reviewers.

Ping: @renyou @goderbauer @Hixie @Piinks Any updates on @flar s queries?

**Comment by Hixie** [comment]
Should pushBackdropFilter take a Paint? Then this would be configurable at the framework layer.

Assuming this doesn't break any customer tests (incl Google tests) then LGTM. If it does break those tests then we'd need to do the migration guide and so forth.

**Comment by flar** [comment]
Having it take a Paint() would introduce a whole bunch of rendering attributes that would raise questions. Perhaps just taking a BlendMode directly since that is the behavior I'm changing would be a more focused modification.

**Comment by Hixie** [comment]
I don't feel strongly on that front, though if Skia is already taking a Paint (it's just a savelayer, right?) then it seems like those questions should already be answered, which would make it a non-issue for us.

**Comment by chinmaygarde** [comment]
Looks like @flar has reservations with adding the paint directly and @Hixie doesn't feel too strongly about it. Based [on this comment](https://github.com/flutter/engine/pull/19631#issuecomment-773688647), it seems like we can land this as is?

**Comment by chinmaygarde** [comment]
There are a couple of analyzer failures. Once those are addressed, lets land this.

**Comment by flar** [comment]
Current status - it is complaining because I didn't implement it on web.

I was able to implement it on canvaskit back end, but there is literally nothing I can find to implement this on HTML back end.

I'm looking at ways to do a "saveLayer collapse" so that BDF can drill through save layers rather than provide a blending mode. If the BDF can run its filter on the entire background then the blending mode is moot - and people are complaining that the BDF doesn't blur the background when there are certain kinds of (saveLayer-using) widgets above it.

I'd love to fix the other problem (BDF not seeing through saveLayers) than to add this mechanism that won't work on web - if I can.

**Comment by Hixie** [comment]
cc @ferhatb @yjbanov

**Comment by flar** [comment]
I'm also considering a couple of alternatives to cause BackdropFilter to collapse the saveLayer tree behind it as the source for the blur which would obsolete the need for this blend mode and possibly meet developer expectations more directly.

**Comment by chinmaygarde** [comment]
In the presence of viable alternatives, if we are unsure about this patch, can we close it for followup later? If not, ping @ferhatb @yjbanov

**Comment by ferhatb** [comment]
What we can do on the web is very limited and browsers have only recently.started supporting CSS backdrop filter. Supporting anything outside the spec will likely break html rendering. https://caniuse.com/css-backdrop-filter.

**Comment by chinmaygarde** [comment]
Looks like this is a no-go based on @ferhatb's last comment. Closing this as it is an old patch with no progress likely.

**Comment by flar** [comment]
There are plenty of things we support that don't work on html rendering. If we use that as an absolute veto of every new feature we can't really grow the platform. @ferhatb ?

**Comment by ferhatb** [comment]
@flar , i didn't mean my comment to be taken as veto. However among the many things we didn't support Shadermasks, advanced gradient shader tilemode, ColorFilter etc, we've been able to close those gaps in most areas. I was commenting on BackdropFilter in particular. Given how inefficient this is in general on any platform, is it not better to give users a better pattern here ? Is this important enough that you want all users to provide if (kIsWeb) buildBlurLayer() else Backdropfilter etc...

**Comment by flar** [comment]
@ferhatb what pattern would you suggest we provide that can work on all platforms?

Right now BDF has surprising behavior in a few cases. Are you saying we should leave it broken so that people will migrate to "a better pattern" that we have neither provided nor outlined?

**Comment by chinmaygarde** [comment]
This looks like a conversation we should have in person to figure out the direction. @flar @Hixie @ferhatb How does next week sound?

**Comment by Hixie** [comment]
Feel free to drop something on my calendar.

**Comment by ferhatb** [comment]
Next week sg. Thanks

**Comment by chinmaygarde** [comment]
The meeting has been scheduled.

**Comment by flar** [comment]
The conclusion seems to be to add a blend mode as a property to the BDF widget. It looks like this PR already includes the necessary plumbing to take the blend mode from the scene builder compositing down to native so all that is missing would be to feed the value from the BDF widget/render object.

**Comment by flar** [comment]
The tests are written, the corresponding change to add a BlendMode property to the BackdropFilter widget is also waiting to be pushed when this engine change is rolled. Everything should be good to go for a review.

@ferhatb @chinmaygarde @Hixie

**Review by chinmaygarde — APPROVED** [review]

**Review by ferhatb — COMMENTED** [review]
LGTM

**Pull request merged**
